File: step_01_basics/1.3_basic_recursion/P006_reverse_an_array/solution.py
"""
Problem: Reverse an Array
Difficulty: EASY | XP: 10

Reverse an array in-place. Show iterative and recursive two-pointer approaches.
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE (New Array -- Extra Space)
# Time: O(n) | Space: O(n)
# ============================================================
def brute_force(arr: List[int]) -> List[int]:
    """Create a new reversed array (not in-place)."""
    n = len(arr)
    return [arr[n - 1 - i] for i in range(n)]


# ============================================================
# APPROACH 2: OPTIMAL (Iterative Two-Pointer Swap)
# Time: O(n) | Space: O(1)
# ============================================================
def optimal(arr: List[int]) -> None:
    """Two pointers from both ends, swap and converge. In-place."""
    left, right = 0, len(arr) - 1
    while left < right:
        arr[left], arr[right] = arr[right], arr[left]
        left += 1
        right -= 1


# ============================================================
# APPROACH 3: BEST (Recursive Two-Pointer Swap)
# Time: O(n) | Space: O(n/2) = O(n) recursion stack
# ============================================================
def best(arr: List[int]) -> None:
    """Recursive version of two-pointer swap. In-place."""
    def helper(left: int, right: int) -> None:
        if left >= right:
            return  # base case: pointers crossed or met
        arr[left], arr[right] = arr[right], arr[left]
        helper(left + 1, right - 1)

    helper(0, len(arr) - 1)


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Reverse an Array ===\n")

    test_cases = [
        ([1, 2, 3, 4, 5], [5, 4, 3, 2, 1]),
        ([1, 2, 3, 4],    [4, 3, 2, 1]),
        ([1],              [1]),
        ([],               []),
        ([1, 2],           [2, 1]),
        ([-3, 0, 5, -1, 7], [7, -1, 5, 0, -3]),
    ]

    for original, expected in test_cases:
        # Brute force (returns new array)
        b = brute_force(original[:])

        # Optimal (in-place)
        o_arr = original[:]
        optimal(o_arr)

        # Best (recursive in-place)
        r_arr = original[:]
        best(r_arr)

        status = "PASS" if b == expected and o_arr == expected and r_arr == expected else "FAIL"
        print(f"Input:     {original}")
        print(f"  Brute:     {b}")
        print(f"  Iterative: {o_arr}")
        print(f"  Recursive: {r_arr}")
        print(f"  Expected:  {expected}  [{status}]\n")

File: step_01_basics/1.3_basic_recursion/P006_reverse_an_array/Solution.java
/**
 * Problem: Reverse an Array
 * Difficulty: EASY | XP: 10
 *
 * Reverse an array in-place. Show iterative and recursive two-pointer approaches.
 *
 * @author DSA_Nova
 */
import java.util.Arrays;

// ============================================================
// Approach 1: Brute Force (New Array -- Extra Space)
// Time: O(n) | Space: O(n)
// ============================================================
class BruteForce {
    /**
     * Create a new array, fill it backwards, then copy back.
     */
    public static int[] reverse(int[] arr) {
        int n = arr.length;
        int[] reversed = new int[n];
        for (int i = 0; i < n; i++) {
            reversed[i] = arr[n - 1 - i];
        }
        return reversed;
    }
}

// ============================================================
// Approach 2: Optimal (Iterative Two-Pointer Swap)
// Time: O(n) | Space: O(1)
// ============================================================
class Optimal {
    /**
     * Two pointers from both ends, swap and converge.
     * Modifies array in-place.
     */
    public static void reverse(int[] arr) {
        int left = 0, right = arr.length - 1;
        while (left < right) {
            // Swap
            int temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
            right--;
        }
    }
}

// ============================================================
// Approach 3: Best (Recursive Two-Pointer Swap)
// Time: O(n) | Space: O(n/2) = O(n) recursion stack
// ============================================================
class Best {
    /**
     * Recursive version of two-pointer swap.
     * Base case: left >= right (pointers crossed or met).
     */
    public static void reverse(int[] arr) {
        helper(arr, 0, arr.length - 1);
    }

    private static void helper(int[] arr, int left, int right) {
        if (left >= right) return; // base case

        // Swap
        int temp = arr[left];
        arr[left] = arr[right];
        arr[right] = temp;

        // Recurse on inner subarray
        helper(arr, left + 1, right - 1);
    }
}

// Main driver
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Reverse an Array ===\n");

        int[][] inputs = {
            {1, 2, 3, 4, 5},
            {1, 2, 3, 4},
            {1},
            {},
            {1, 2},
            {-3, 0, 5, -1, 7},
        };
        int[][] expected = {
            {5, 4, 3, 2, 1},
            {4, 3, 2, 1},
            {1},
            {},
            {2, 1},
            {7, -1, 5, 0, -3},
        };

        for (int t = 0; t < inputs.length; t++) {
            // Brute force (returns new array)
            int[] b = BruteForce.reverse(inputs[t].clone());

            // Optimal (in-place)
            int[] oArr = inputs[t].clone();
            Optimal.reverse(oArr);

            // Best (recursive in-place)
            int[] rArr = inputs[t].clone();
            Best.reverse(rArr);

            boolean pass = Arrays.equals(b, expected[t])
                        && Arrays.equals(oArr, expected[t])
                        && Arrays.equals(rArr, expected[t]);

            System.out.println("Input:     " + Arrays.toString(inputs[t]));
            System.out.println("Brute:     " + Arrays.toString(b));
            System.out.println("Iterative: " + Arrays.toString(oArr));
            System.out.println("Recursive: " + Arrays.toString(rArr));
            System.out.println("Expected:  " + Arrays.toString(expected[t]));
            System.out.println("Pass:      " + pass);
            System.out.println();
        }
    }
}
